proto/gossip_delegates.go

```diff
@@ -201,18 +227,52 @@ func (gd *GossipDelegate) NotifyLeave(node *memberlist.Node) {
 	if nodeName == gd.nodeId {
 		gd.triggerStateEvent(types.SELF_LEAVE)
 	} else {
-		err := gd.UpdateNodeStatus(types.NodeId(nodeName), types.NODE_STATUS_DOWN)
-		if err != nil {
-			logrus.Infof("gossip: Could not update status on NotifyLeave : %v", err.Error())
-			return
+		if gd.quorumProvider.Type() == types.QUORUM_PROVIDER_FAILURE_DOMAINS {
+			go func() {
+				isSuspect := gd.isNodeSuspect(types.NodeId(nodeName))
+				if isSuspect {
+					gd.setNodeAsSuspectOffline(nodeName)
+				} else {
+					gd.setNodeOffline(nodeName)
+				}
+			}()
+		} else {
+			gd.setNodeOffline(nodeName)
 		}
-		gd.triggerStateEvent(types.NODE_LEAVE)
 	}
 
 	gd.updateGossipTs()
 	return
 }
 
+func (gd *GossipDelegate) setNodeAsSuspectOffline(nodeName string) {
+	err := gd.UpdateNodeStatus(types.NodeId(nodeName), types.NODE_STATUS_SUSPECT_DOWN)
+	if err != nil {
+		logrus.Infof("gossip: Could not update status on NotifyLeave : %v", err.Error())
+		return
+	}
+	logrus.Infof("gossip: Node %v is suspected offline", nodeName)
+	gd.triggerStateEvent(types.NODE_LEAVE)
+	// Add the node to probation list
+	if ok := gd.nodeDownProbationManager.Exists(gd.nodeNameToProbationID(nodeName)); ok {
+		logrus.Infof("gossip: Node %v already exists in probation list. ", nodeName)
+	} else {
+		if err := gd.nodeDownProbationManager.Add(gd.nodeNameToProbationID(nodeName), nil, false); err != nil {
+			logrus.Warnf("gossip: Unable to add suspected down node %v to probation list: %v", nodeName, err)
+		}
+		logrus.Infof("gossip: Node %v added to probation list", nodeName)
+	}
+}
+
+func (gd *GossipDelegate) setNodeOffline(nodeName string) {
+	err := gd.UpdateNodeStatus(types.NodeId(nodeName), types.NODE_STATUS_DOWN)
+	if err != nil {
```

Comment:
UpdateNodeStatus returns an error only if the node for which the update was done, does not exist in our store. We don't return here, because the caller of this function is a memberlist callback function which is predefined and does not return an error.